lib/h3.rb

```diff
@@ -124,4 +130,85 @@ def self.h3_unidirectional_edges_from_hexagon(origin)
     getH3UnidirectionalEdgesFromHexagon(origin, edges)
     edges.read_array_of_ulong_long(max_edges).reject { |i| i == 0 }
   end
+
+  def self.h3_to_geo_boundary(h3_index)
+    geo_boundary = Structs::GeoBoundary.new
+    h3ToGeoBoundary(h3_index, geo_boundary)
+    geo_boundary[:verts].take(geo_boundary[:num_verts] * 2).map do |d|
+      rads_to_degs(d)
+    end.each_slice(2).to_a
+  end
+
+  def self.hex_ranges(h3_set, k)
+    h3_set.uniq!
+    max_out_size = h3_set.size * max_kring_size(k)
+    out = FFI::MemoryPointer.new(H3_INDEX, max_out_size)
+    pentagonal_distortion = false
+    FFI::MemoryPointer.new(H3_INDEX, h3_set.size) do |h3_set_ptr|
+      h3_set_ptr.write_array_of_ulong_long(h3_set)
+      pentagonal_distortion = hexRanges(h3_set_ptr, h3_set.size, k, out)
+    end
+    raise(ArgumentError, "One of the specified hexagon ranges contains a pentagon") if pentagonal_distortion
+
+    h3_range_indexes = out.read_array_of_ulong_long(max_out_size)
+    out = h3_set.inject({}) { |acc, i| acc[i] = []; acc }
+    h3_set.each_with_index do |h3_index, i|
+      (k + 1).times { out[h3_index] << [] }
+      0.upto(h3_set.count * max_kring_size(k) / h3_set.count).map do |j|
+        ring_index = ((1 + Math.sqrt(1 + 8 * (j / 6.0).ceil)) / 2).floor - 1
```

Comment:
Just to know... where does this logic come from?

Reply:
This one is surprising to me.

I don't know the details, but I guess a C program calling this C lib gets the output digested and ready to be used. Why isn't our code (or the Python binding) just building a straightworward Ruby structure from the output as it happens with other wrapped methods?

Not saying it does not have to be done, just wanting to understand what makes this different.

Reply:
The output is a single array of h3 indexes. If the k distance is quite large, then the array is quite long and the boundaries between hex ranges aren't simple to determine.

I tried a few things myself and then went with the py bindings approach since it seems to work.

Reply:
My point is, again, about parity.

Put yourself on the shoes of someone that knows H3: They see `h3_ranges` which by convention wraps `h3Ranges`, wouldn't they therefore expect the same result?

Reply:
@dfellis then, all C client code is repeating logic like

```ruby
((1 + Math.sqrt(1 + 8 * (j / 6.0).ceil)) / 2).floor
```

? That is what seems more suspicious to me, I'd be surprised if that was the case. On the other hand, [the Go port](https://github.com/uber/h3-go/blob/7cdeef8ae31e1cefc73c6ccb32e653b06315f15f/h3.go#L206-L224) looks simpler.

Unfortunately I am not familiar with H3, so can only reason in a formal way. Which is limited and have questions but can't anticipate answers :).

Your experience/feedback is going to be very helpful for us to help understand the rationale behind this implementation and move forward.

Reply:
OK!

Reply:
Thanks @fxn and @dfellis 🙌 

I've created #24 to return to the discussion of the best behaviour for this method.